prov/util/src/util_mr_cache.c

```diff
@@ -125,21 +125,42 @@ static void util_mr_uncache_entry_storage(struct ofi_mr_cache *cache,
 }
 
 static void util_mr_uncache_entry(struct ofi_mr_cache *cache,
-				  struct ofi_mr_entry *entry)
+				  struct ofi_mr_entry *entry,
+				  int defer)
 {
 	util_mr_uncache_entry_storage(cache, entry);
 
 	if (entry->use_cnt == 0) {
 		dlist_remove(&entry->list_entry);
-		dlist_insert_tail(&entry->list_entry, &cache->flush_list);
+		if (defer) {
+			dlist_insert_tail(&entry->list_entry, &cache->flush_list);
+		} else {
+			/* This is safe to unlock since the mm_lock
+			 * only protects the cache storage which is not
+			 * accessed in util_mr_free_entry. It is necessary
+			 * to unlock before util_mr_free_entry because the
+			 * mm_lock needs to be unlocked before a possible
+			 * monitor triggering event (such as a free or
+			 * deregistration) or else there will be a
+			 * deadlock (The monitor will try to obtain this
+			 * lock before calling ofi_mr_cache_notify).
+			 */
+			pthread_mutex_unlock(&mm_lock);
+			util_mr_free_entry(cache, entry);
+			pthread_mutex_lock(&mm_lock);
+		}
```

Comment:
In the non-deferred case, can we just have the monitor call flush immediately after the notify call?  That might be cleaner.

Reply:
Hmm I think so